tools/distributed/pytorch_dist_exec.py

```diff
@@ -0,0 +1,109 @@
+import subprocess
+import os
+import socket
+from argparse import ArgumentParser, REMAINDER
+
+import torch
+
+def parse_args():
+    """
+    Helper function parsing the command line options
+    @retval ArgumentParser
+    """
+    parser = ArgumentParser(description="PyTorch Exec is a helper utiliy that "
+            "spawns up multiple distributed training processes. The utility "
+            "can be used in single-node distributed training that each "
+            "distributed process will be operating on a single GPU. (for "
+            "well-improved performance reasons). "
+            "In this case, this utilily will launch a given number of "
+            "processes per node (nproc_per_node) , while this number needs to "
+            "be smaller than the number of GPUs (n_gpus) on the current system,"
+            " and each process will be operating on a single GPU from GPU 0 to "
+            "GPU nproc_per_node - 1. "
+
+            "This utility can be further used for multi-node "
+            "distributed training by spawning up multiple processes per node "
+            "for well-improved distributed performance as well. This will "
+            "especially be benefitial for systems with multiple Infiniband "
+            "interfaces since all of them can be utilized for aggregated "
+            "communication bandwidth. Please note that this utilty and "
+            "multi-process/node distributed single node or multi-node "
+            "training currently only supports the NCCL distributed backend. "
+            "This utilty helper will require that training script is able to "
+            "parse --device=X as an argument since it will be injected by this "
+            "utility. "
```

Comment:
Why can't we set `CUDA_VISIBLE_DEVICES`? This is usually what you want, because it will also ensure that the processes won't ever share a GPU, no matter what your training script does

Reply:
I was initially using CUDA_VISIBLE_DEVICES and found NCCL doesn't work for this purpose and had to switch to something else.

Synced with NVIDIA NCCL folks,  CUDA IPC needs to see all the devices, that is the reason why NCCL doesn't support this.  Looks like the only known way (according to NCCL team) is to set the device using cudaSetDevice in the code :|